### PR TITLE
Add support for profiles that are configured for MFA

### LINKED ISSUE #141: Add support for profiles that require MFA
**Is your feature request related to a problem? Please describe.**
We are currently not handling the Multi Factor Authentication callback for user profiles that require MFA.
We need to add support for that.

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

---

This is a :rocket: feature request

### PR DESCRIPTION
**Pull request by normj**
*Description of changes:*
Hook up a callback to the AssumeRoleAWSCredentials object for the user to enter an MFA token. This is a port for the MFA support in Amazon.Lambda.Tools.

MFAExperience

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

Closes #141

### CODE REVIEW

File: test/AWS.Deploy.CLI.UnitTests/TestToolInteractiveServiceImpl.cs [hunk 1]
```diff
@@ -14,7 +14,7 @@ public class TestToolInteractiveServiceImpl : IToolInteractiveService
 
         private IList<string> InputCommands { get; set; }
 
-        public TestToolInteractiveServiceImpl(): this(new List<string>())
+        public TestToolInteractiveServiceImpl() : this(new List<string>())
```
Comment (ppittle) [inline test/AWS.Deploy.CLI.UnitTests/TestToolInteractiveServiceImpl.cs:17]: nit: unnecessary whitespace change

**Review by ppittle — APPROVED** [review]

File: src/AWS.Deploy.CLI/AWSUtilities.cs [hunk 2]
```diff
@@ -22,59 +23,75 @@ public AWSUtilities(IToolInteractiveService toolInteractiveService)
...
-                if (chain.TryGetAWSCredentials(profileName, out credentials) &&
+                AWSCredentials credentials;
+
+                var chain = new CredentialProfileStoreChain();
+
+                if (!string.IsNullOrEmpty(profileName))
+                {
+                    if (chain.TryGetAWSCredentials(profileName, out credentials) &&
+                        // Skip checking CanLoadCredentials for AssumeRoleAWSCredentials because it might require an MFA token and the callback hasn't been setup yet.
+                        (credentials is AssumeRoleAWSCredentials || await CanLoadCredentials(credentials)))
```
Comment (ganeshnj) [inline src/AWS.Deploy.CLI/AWSUtilities.cs:38]: I think you can combine the nested ifs to one like you did below L45.

File: src/AWS.Deploy.CLI/AWSUtilities.cs [hunk 3]
```diff
@@ -22,59 +23,75 @@ public AWSUtilities(IToolInteractiveService toolInteractiveService)
...
         {
-            AWSCredentials credentials;
 
-            var chain = new CredentialProfileStoreChain();
 
-            if (!string.IsNullOrEmpty(profileName))
+            async Task<AWSCredentials> Resolve()
             {
-                if (chain.TryGetAWSCredentials(profileName, out credentials) &&
+                AWSCredentials credentials;
```
Comment (ganeshnj) [inline src/AWS.Deploy.CLI/AWSUtilities.cs:30]: I found it confusing how many places the same instance of credentials is used. Can we make it context specific.
```csharp
    var chain = new CredentialProfileStoreChain();

    if (!string.IsNullOrEmpty(profileName))
    {
        if (chain.TryGetAWSCredentials(profileName, out var profileCredentials) &&
            // Skip checking CanLoadCredentials for AssumeRoleAWSCredentials because it might require an MFA token and the callback hasn't been setup yet.
            (profileCredentials is AssumeRoleAWSCredentials || await CanLoadCredentials(profileCredentials)))
        {
            _toolInteractiveService.WriteLine($"Configuring AWS Credentials from Profile {profileName}.");
            return profileCredentials;
        }
    }

    if (!string.IsNullOrEmpty(lastUsedProfileName) &&
        chain.TryGetAWSCredentials(lastUsedProfileName, out var lastUsedCredentials) &&
        await CanLoadCredentials(lastUsedCredentials))
    {
        _toolInteractiveService.WriteLine($"Configuring AWS Credentials with previous configured profile value {lastUsedProfileName}.");
        return lastUsedCredentials;
    }

    try
    {
        var fallbackCredentials = FallbackCredentialsFactory.GetCredentials();

        if (await CanLoadCredentials(fallbackCredentials))
        {
            _toolInteractiveService.WriteLine("Configuring AWS Credentials using AWS SDK credential search.");
            return fallbackCredentials;
        }
    }
    catch (AmazonServiceException)
    {
        // FallbackCredentialsFactory throws an exception if no credentials are found. Burying exception because if no credentials are found
        // we want to continue and ask the user to select a profile.
    }

    var sharedCredentials = new SharedCredentialsFile();
    if (sharedCredentials.ListProfileNames().Count == 0)
    {
        _toolInteractiveService.WriteErrorLine("Unable to resolve AWS credentials to access AWS.");
        throw new NoAWSCredentialsFoundException();
    }

    var consoleUtilities = new ConsoleUtilities(_toolInteractiveService);
    var selectedProfileName = consoleUtilities.AskUserToChoose(sharedCredentials.ListProfileNames(), "Select AWS Credentials Profile", null);

    if (!chain.TryGetAWSCredentials(selectedProfileName, out var selectedProfileCredentials) ||
        !(await CanLoadCredentials(selectedProfileCredentials)))
    {
        _toolInteractiveService.WriteErrorLine($"Unable to create AWS credentials for profile {selectedProfileName}.");
        throw new NoAWSCredentialsFoundException();
    }

    return selectedProfileCredentials;
}
```

**Review by ganeshnj — CHANGES_REQUESTED** [review]

File: src/AWS.Deploy.CLI/AWSUtilities.cs (hunk 2 above)
Comment (normj) [reply src/AWS.Deploy.CLI/AWSUtilities.cs:38]: Done

File: test/AWS.Deploy.CLI.UnitTests/TestToolInteractiveServiceImpl.cs (hunk 1 above)
Comment (normj) [reply test/AWS.Deploy.CLI.UnitTests/TestToolInteractiveServiceImpl.cs:17]: Done

File: src/AWS.Deploy.CLI/AWSUtilities.cs (hunk 3 above)
Comment (normj) [reply src/AWS.Deploy.CLI/AWSUtilities.cs:30]: Done

**Review by ganeshnj — APPROVED** [review]

**Review by ganeshnj — APPROVED** [review]
